keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+
+An example if using security profile is
+
+```
+kubeadm init --security-profile=default-1.0-0
+```
+
+The command above will configure a secure cluster according to security profile `default-1.0-0`.
+
+As security profiles cover common/generic use cases,
+they are portable and reusable in most Kubernetes deployments,
+They are **NOT** defining policies specific to a particular Kubernetes deployment, like:
+
+- ResourceQuota
+- RBAC rules
+- Rules in NetworkPolicy for particular workloads
+
+The proposal also works with Config-as-Code systems,
+see the [section](#work-with-config-as-code-systems) below for details.
+
+### Proposed Security Profiles
+
+#### default
+
+Full example: [default-1.0-0](#the-default-profile)
+
+Minimum security practice to be compatible with most workloads.
+Best fit for sharing the cluster with trusted workloads.
+
+Enforcement:
+
+- etcd
+  - TLS required for communication
+  - Explicitly specify `--wal-dir`, and `--max-wals=0`
+- kube-apiserver
+  - TLS required for communication
+  - Audit logging must be explicitly configured
+  - No anonymous auth, basic auth (unset `--anonymous-auth`, `--basic-auth-file`)
+  - Required authorization-mode: Node, RBAC
+  - Excluded authorization-mode: ABAC, AlwaysAllow
+  - Disable insecure port (`--insecure-port=0`, unset `--insecure-bind-address`)
+  - No auth token file (unset `--token-auth-file`)
+  - Disable profiling (unset `--profiling`)
+  - Request timeout is explicitly set
+  - Use service account credentials (`--use-service-account-credentials=true`)
+  - Required admission controls: AlwaysPullImages, DenyEscalatingExec, NamespaceLifecycle, ResourceQuota, PodSecurityPolicy, ServiceAccount, NodeRestriction, EventRateLimit
+  - Excluded admission controls: AlwaysAdmit
+  - Required feature gates to be ON: AdvancedAuditing, RotateKubeletServerCertificate, RotateKubeletClientCertificate
+- controller-manager
+  - Explicitly set `--terminated-pod-gc-threshold`
+- kubelet
+  - Disable hostname override (unset `--hostname-override`)
+  - No privileged
+- addons
+  - No dashboard
+- file-permissions
+  - ... # correct setup of cert/key/config file permissions
+- PodSecurityPolicy
+  - No priviledged
+- NodeTaints
+  - No schedule on master
+
+#### saas-multitenancy
+
+Full Example: [saas-multitenancy-1.0-0](#the-saas-multitenancy-profile)
+
+For SaaS scenarios which allows only plain workloads without the need to gain privileged access.
+
+Enforcement: in addition to `default`
+
+- PodSecurityPolicy
+  - No privileged pods
+  - Not allow privilege escalation
+  - AppArmor profiles: docker/default, runtime/default
+  - Seccomp profiles: docker/default, runtime/default
+  - Drop ALL capabilities
+  - No host namespaces (network, IPC, PID)
+  - Limit volume types to
+    - configMap
+    - downwardAPI
+    - emptyDir
+    - nfs
+    - persistentVolumeClaim
+    - projected
+    - secret
+- NetworkPolicy
+  - Block all ingress by default
+  - Block all egress by default
+
```

Comment:
Please add secrets encryption

Reply:
I will add that in bootstrapping rules via requirement of --experimental-encryption-provider-config